src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,144 @@
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index  "
+                    + index + "is outside the list");
+        }
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index  "
+                    + index + "is outside the list");
+        }
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index  " + index + "is outside the list");
+        }
+        final T value = elements[index];
+        T[] buffer = (T[]) new Object[elements.length];
+        rewriteArray(elements,buffer,index,0);
+        rewriteArray(buffer,elements,1,index);
+        size--;
+        return value;
     }
 
     @Override
-    public T remove(T element) {
-        return null;
+    public T remove(T element) throws NoSuchElementException {
+        if (isEmpty()) {
+            throw new ArrayListIndexOutOfBoundsException("Can't remove element from empty list");
+        }
+        int index;
+        for (index = 0; index < elements.length; index++) {
+            if (elements[index] == element
+                    || elements[index] != null && elements[index].equals(element)) {
+                return remove(index);
+            }
+        }
+        throw new NoSuchElementException();
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private int getIndexFreeSell() {
+        size = elements.length;
+        growthElementsArray();
+        return size;
+    }
+
+    private T [] growthElementsArray() {
+        int newLength = elements.length + elements.length / 2;
+        T[] buffer = (T[]) new Object[elements.length];
+        rewriteArray(elements, buffer, 0,0);
+        elements = (T[]) new Object[newLength];
+        rewriteArray(buffer, elements, 0,0);
+        return elements;
+    }
+
+    private void rewriteArray(T[] in, T[] out, int startIn,int startOut) {
+        int j = startOut;
+        int i;
+        for (i = startIn; i < in.length; i++) {
+            if (j < out.length) {
+                out[j] = in[i];
+                j++;
+            }
+        }
+    }
+
+    private void releaseCell(int index) {
+        T[] buffer = (T[]) new Object[elements.length - index + 1];
+        rewriteArray(elements,buffer,index,0);
+        rewriteArray(buffer,elements,0,index + 1);
     }
```

Comment:
<img width="572" alt="image" src="https://github.com/mate-academy/jv-array-list/assets/81781479/bced0183-b0b9-472e-9466-1abe7933cc73">

you can move elements from and to the same array